### ISSUE TITLE
How to handle ETERM?

### ISSUE THREAD

**Issue by haf-raysearchlabs** [zeromq/fszmq#134]
What are the patterns for dealing with ETERM and cancellation?

Also, a few operations say they might throw "TimeoutException"; under what circumstances do they and is there a way to avoid that? It would seem the `tryRecv` function, for example, is harder to use (requires a buffer, index into the buffer) and doesn't return the same level of typing as `recv`.

I'm looking to create a proxy and a many-senders to many receivers; as such: https://github.com/logary/logary/blob/feature/log-router/src/services/Logary.Services.Rutta/Rutta.fs

**Comment by pblasucci** [comment]
Good questions...

The best approach for handling cancellation will depend on the specifics of your application. So, I'm afraid I'd need more info to answer that question.

As for `TimeoutException`, that will only be thrown from a non-polling send or receive if, and only if, the appropriate socket option is set (e.g. `ZMQ.RCVTIMEO`, `ZMQ.SNDTIMEO`). It's primarily meant to be used with non-polling, blocking socket operations. Incidentally, the operations you'll normally want to use are either polling-based callbacks (see the `Polling` module... especially `tryPollInput`) or the blocking operations on the `Socket` module ('send`, `sendMore`, `recvAll`, et cetera). Meanwhile `trySend` and `tryRecv` are provided for more-exotic scenarios. In particular, `tryRecv` is meant to be a performance optimization for scenarios where you have fixed-length (known ahead-of-time) messages. Non-blocking uses of `Socket.trySend` are, in my experience, rare. But they're potentially useful for identify stalled pipelines.

As for your overall goals, does the proxy do anything specific before forwarding messages? If not, the out-of-the-box ZMQ proxy may be of interest to you (see the `Proxying` module). For building out many-senders/many-receivers, it seems you're on the right track. But I'm confused as to what you hope to gain from `tryRecv` (as opposed to `recv` or, ideally, `recvAll`)?

I'm sorry I don't have a better answer for you at this time. But I'm happy to discuss this further.

**Comment by haf-raysearchlabs** [comment]
> As for TimeoutException, that will only be thrown from a non-polling send or receive if, and only if, the appropriate socket option is set (e.g. ZMQ.RCVTIMEO, ZMQ.SNDTIMEO)

Is it set by default? It sounds like it's not.

Does `recv` then return `null` if it's cancelled?

My aim was to avoid copying all data into memory if I'm just proxying; does `Proxying.proxy` solve that? I'm also thinking of taking Suave's BufferManager and feed it into this lib to reuse the memory, which doesn't seem to be the case for `recvAll`, right?

> does the proxy do anything specific before forwarding messages.

No.

The `router` deserialise them and feed them into Logary.

The pattern is thought like: `sender 1..* -> proxy -> router`, like [described here](https://github.com/logary/logary/issues/97). Any sender should not block proxy or router. Proxies should provide backpressure from blocking routers. Logary contains internal buffers that let us block when ingestion/creation of messages is at a higher rate than we can send them to the infrastructure (influx, elasticsearch).

In summary; right to have a XSUB and a XPUB in proxy which is linked up with `Proxying.proxy`? And to have a `SUB` in `router` that gets messages from XPUB in proxy?

**Comment by pblasucci** [comment]
`ZMQ.RCVTIMEO` and `ZMQ.SNDTIMEO` are not set by default (technically, they're set to "inifinity"... but it amounts to the same thing).

Currently, any socket or poller will raise an exception if it has an operation pending and `ETERM` is signaled. I've long wanted to revisit this and handle it more safely (possibly in the next release). However, the current work-around is to provide your own `try ... with`.

`Proxying.proxy` never pulls the message into _managed space_ (i.e. CLR or Mono). As for `recvAll`, it delegates to `recv`, which marshals each message frame (unmanaged memory) into a new byte array (managed memory). Though I'm not sure I fully understand your concerns about memory reuse.

Overall, though, your pattern sound right. `XSUB`/`XPUB` proxy with a `SUB` router makes sense.

**Comment by pblasucci** [comment]
One other thing I forgot to mention. The samples on the zguide are (very slightly) out-of-date. An up-to-date and more complete set of examples may be found here: 

http://zeromq.github.io/fszmq/zguide/index.html

Included therein is a primitive (but workable) example of [handling program termination](http://zeromq.github.io/fszmq/zguide/interrupt.html)

_(Aside: I've personally adapted it to work with `CancellationToken`s on many projects and found it to be quite effective.)_ 

Hopefully this is of some use to you.

**Comment by pblasucci** [comment]
@haf Where'd you ever wind up with this? Is there more conversation to be had (or work to be done)? Should we close it and move on?

**Comment by haf** [comment]
@pblasucci Well; could it be a documentation issue? It would be great if the zguide was updated to be relevant to this API, for example.

**Comment by pblasucci** [comment]
OK. I'm all for better documentation. Though, at this point, I wouldn't likely update the zguide (as it is across-the-board out-of-date). However, we can certainly continue to buuild out the docs in the fszmq gh-pages site. After I put out something about the new Options API, I will add something which addresses: 

  + blocking sned/recv (with `TimeoutException`)
  + handling `EINTR` in non-blocking code
  + clean application shutdown

Does that cover the areas of concern? Is there some specific scenarios you'd especially like to see covered?

Thanks.

**Comment by haf** [comment]
Yes, the exceptions and shutdown cases are rather important and very easy to miss without docs. I'm not even sure I handle it correctly in Rutta/Shipper in Logary.

**Comment by pblasucci** [comment]
Created new issue ( #144 ) to track some of the action items that arose from this conversation.
